### DISCUSSION TITLE
Admin API /users/ POST request data

### DISCUSSION THREAD

**Discussion by pcnoic** [fief-dev/fief#285] [Q&A] [answered]
Hey folks, I am working with the Admin API in order to create users programmatically from my own API, after a certain webhook is triggered. Unfortunately, I cannot find what the proper data are for POSTing to `/users/`. The OpenAPI documentation on `/docs` only describes the expected responses from the specific endpoint. 

I though of inspecting the request that is sent from the fief dashboard and I ended up with something like this:

```
user_data = {
            "email": 
            "password": secrets.token_hex(16),
            "tenant": 
            "query":
            "fields-name": 
            "fields-address.line1": 
            "fields-address.line2": 
            "fields-address.postal_code": 
            "fields-address.city": 
            "fields-address.state": 
            "fields-address.country": 
            "fields-bank_account":
            "fields-bank_account_owner": 
            "fields-mobile": 
        }
```

Even with these data, I get  `422 Unprocessable Entity` responses.

In the above, the `csrf_token` that is sent from the fief dashboard UI is missing. I am not 100% sure if it's required or how to generate one from my own API. I would appreciate some help on how to determine the correct `data` for the request.

**Comment by pcnoic** [comment]
After some tinkering with a REST client, I figured out that the request data should indeed be in JSON format.

```
{
    "email": "reducted@reducted.com",
    "password": "reducted",
    "email_verified": true,
    "tenant_id": "reducted",
}
```

This will indeed register a user programmatically. Still working on finding out whether's possible to POST custom user fields.

**Comment by pcnoic** [comment, marked as answer]
My bad. Essentially what is returned from the POST endpoints, are values you can also pass to the request as data. So, for anything you want to add to user fields, you have to use `fields: {}`.

https://example.fief.dev/admin/api/docs#/Users/users_update_users__id__patch

Leaving this here for anyone that might have the same (naive) question as me in the future. Cheers.

**Reply by Bananen1234** [reply]
Hi! So the payload should look like this: { fields: { email: a.a@a.com, password: abc123 } 
}

Or?

**Comment by frankie567** [comment]
Hi @pcnoic 👋 

Thanks for raising this point, I didn't notice OpenAPI was not generated correctly for this endpoint. Opening a issue to fix this!

**Reply by Bananen1234** [reply]
How is this going? Because right now theres no explanation.

**Reply by frankie567** [reply]
It looks like this:

```json
{
  "email": "user@example.com",
  "email_verified": true,
  "is_active": true,
  "tenant_id": "string",
  "fields": {
    "my_custom_field": "foo"
  }
}
```
